src/core/search/block_list.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <absl/types/span.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <optional>
+#include <vector>
+
+#include "core/search/base.h"
+#include "core/search/compressed_sorted_set.h"
+
+namespace dfly::search {
+// BlockList is a container wrapper for CompressedSortedSet / vector<DocId>
+// to divide the full sorted id range into separate blocks. This reduces modification
+// complexity from O(N) to O(logN + K), where K is the max bock size.
+//
+// It tires to balance block sizes in the range [block_size / 2, block_size * 2]
+// by splitting or merging nodes when needed.
+template <typename C /* underlying container */> class BlockList {
+  using BlockIt = typename PMR_NS::vector<C>::iterator;
+  using ConstBlockIt = typename PMR_NS::vector<C>::const_iterator;
+
+ public:
+  BlockList(PMR_NS::memory_resource* mr, size_t block_size = 1000)
+      : block_size_{block_size}, blocks_(mr) {
+  }
+
+  // Insert element, returns true if inserted, false if already present.
+  bool Insert(DocId t);
+
+  // Remove element, returns true if removed, false if not found.
+  bool Remove(DocId t);
+
+  size_t Size() const {
+    return size_;
+  }
+
+  size_t size() const {
+    return size_;
+  }
+
+  struct BlockListIterator {
+    // To make it work with std container contructors
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = DocId;
+    using pointer = DocId*;
+    using reference = DocId&;
+
+    DocId operator*() const {
+      return **block_it;
+    }
+
+    BlockListIterator& operator++();
+
+    friend class BlockList;
+
+    friend bool operator==(const BlockListIterator& l, const BlockListIterator& r) {
+      return l.block_it == r.block_it;
+    }
+
+    friend bool operator!=(const BlockListIterator& l, const BlockListIterator& r) {
+      return !operator==(l, r);
+    }
+
+   private:
+    BlockListIterator(ConstBlockIt begin, ConstBlockIt end) : it(begin), it_end(end) {
+      if (it != it_end) {
+        block_it = it->begin();
+        block_end = it->end();
+      }
+    }
+
+    ConstBlockIt it, it_end;
+    std::optional<typename C::iterator> block_it, block_end;
+  };
+
+  BlockListIterator begin() const {
+    return BlockListIterator{blocks_.begin(), blocks_.end()};
+  }
+
+  BlockListIterator end() const {
+    return BlockListIterator{blocks_.end(), blocks_.end()};
+  }
+
+ private:
+  // Find block that should contain t. Returns end() only if empty
+  BlockIt FindBlock(DocId t);
+
+  // If needed, try merging with previous block
+  void TryMerge(BlockIt block);
+
+  // If needed, try splitting
+  void TrySplit(BlockIt block);
```

Comment:
```suggestion
  void MergeWithPrevBlockIfNeeded(BlockIt block);

  void SplitIfNeeded(BlockIt block);
```

Reply:
I thin the Try already shows it's not a 100% operation, and I like having a small comment more than a long clumsy name... TryMerging, what happens inside is implementation detail 🙂